### ISSUE TITLE
`fft` and `ifft` call `vDSP_create_fftsetupD` without calling the corresponding destroy function, resulting in memory leaks

### ISSUE THREAD

**Issue by harenbrs** [scottsievert/swix#9]
In `ndarray/complex-math.swift`, `fft(x: ndarray)` and `ifft(yr: ndarray, yi: ndarray)` each call `vDSP_create_fftsetupD` to set up the following call of `vDSP_fft_zropD`. The system then allocates memory that is never released.

Simply appending a call to `vDSP_destroy_fftsetupD` at the end of each function would technically fix the leak, but these functions are not meant to be called in a performance-sensitive context anyway. In my own project, I have split the method into two, the first of which is only called once per value of `N`:

```swift
func fftInit(N: Int) -> FFTSetupD {
    var radix: FFTRadix = FFTRadix(FFT_RADIX2)
    var pass: vDSP_Length = vDSP_Length((log2(N.double)+1.0).int)
    var setup: FFTSetupD = vDSP_create_fftsetupD(pass, radix)
    return setup
}
```

```swift
func fft(x: ndarray, withSetup setup: FFTSetupD) -> (ndarray, ndarray) {
    //... (where the contents of fftInit above have been removed)
    vDSP_fft_zropD(setup, &x2, stride, &y, stride, log2n.length, dir)
    //...
}
```

The non-memory-leaking alternative closest to the original would be:

```swift
func fft(x: ndarray) -> (ndarray, ndarray) {
    //...
    var setup:FFTSetupD = vDSP_create_fftsetupD(pass, radix)
    //...
    vDSP_fft_zropD(setup, &x2, stride, &y, stride, log2n.length, dir)
    vDSP_destroy_fftsetupD(setup)
    //...
}
```

The reason I have not submitted a pull request is that clearly there is a decision to be made here: performance vs flexibility. My project requires performance, so I create the `FFTSetupD` structure once and reuse it over and over. **Swix**, as I understand it, is meant to emulate the **NumPy** model, where the flexibility of being able to use one single function call to compute the FFT is preferred over potential speed gains.

In summary, I would like to draw the maintainer's attention to the memory leak caused by repeated calls of `fft` or `ifft`, and introduced by `vDSP_create_fftsetupD`. I would also like to start a discussion on the limits of emulating the elegance of NumPy in Swift, and making number crunching in Swift fast and flexible at the same time.

**Comment by scottsievert** [comment]
I would be inclined to say by default it should be slow but not leak. If speed is desired, it would be advantageous to create an optional performance parameter. Maybe something like `fft(x, setup:FFTSetup?)` which would require the user to run another function for the setup. This jives with my internal compass; more work is required for more speed (and we should make that work easy).

Before we do that, are we sure that this is a significant performance hit? I'm inclined to believe it's not. Yeah it might take 10 times as long but the FFT is surprisingly fast and there is often other computation to be done. Numpy seems to follow this model; I can't detect any `fft.init` or equivalent anywhere.

To be clear, I would recommend simply inserting conditionals at the beginning and end of the function where the setup and destroy functions go. If the condition is not met and `setup == nil` we could *assume* (and document!) that `fft_setup(N)` (and a destroy function if necessary).

Feel free to submit a pull request for anything -- I can always them down. It could just add the destroy function or include the optional parameter as described. I would implement what you describe but am running short on time and am not incentivized to devote significant time to this project. Please, I will gladly test and test appropriate requests.

**Comment by scottsievert** [comment]
Closed by PR #10 (implemented in PR #11)
